src/plugins/embeddable/public/embeddable_panel/panel_header/use_embeddable_panel_context_menu.tsx

```diff
@@ -0,0 +1,179 @@
+  EuiButtonIcon,
+  EuiContextMenu,
+  EuiContextMenuPanelDescriptor,
+  EuiPopover,
+} from '@elastic/eui';
+import { Action, buildContextMenuForActions } from '@kbn/ui-actions-plugin/public';
+
+import { uiActions } from '../../kibana_services';
+import { EmbeddablePanelProps, PanelUniversalActions } from '../types';
+import { getContextMenuAriaLabel } from '../embeddable_panel_strings';
+import { useSelectFromEmbeddableInput } from '../use_select_from_embeddable';
+import { IEmbeddable, contextMenuTrigger, CONTEXT_MENU_TRIGGER, ViewMode } from '../..';
+
+const sortByOrderField = (
+  { order: orderA }: { order?: number },
+  { order: orderB }: { order?: number }
+) => (orderB || 0) - (orderA || 0);
+
+const removeById =
+  (disabledActions: string[]) =>
+  ({ id }: { id: string }) =>
+    disabledActions.indexOf(id) === -1;
+
+export const useEmbeddablePanelContextMenu = ({
+  index,
+  embeddable,
+  getActions,
+  actionPredicate,
+  universalActions,
+}: {
+  index?: number;
+  embeddable: IEmbeddable;
+  universalActions: PanelUniversalActions;
+  getActions: EmbeddablePanelProps['getActions'];
+  actionPredicate?: (actionId: string) => boolean;
+}) => {
+  const [isContextMenuOpen, setIsContextMenuOpen] = useState(false);
+  const [contextMenuActions, setContextMenuActions] = useState<Array<Action<object>>>([]);
+  const [contextMenuPanels, setContextMenuPanels] = useState<EuiContextMenuPanelDescriptor[]>([]);
+
+  const title = useSelectFromEmbeddableInput('title', embeddable);
+  const viewMode = useSelectFromEmbeddableInput('viewMode', embeddable);
+
+  const mounted = useMountedState();
+
+  const getAllPanelActions = useCallback(async () => {
+    const regularActions = await (async () => {
+      if (getActions) return await getActions(CONTEXT_MENU_TRIGGER, { embeddable });
+      return (
+        (await uiActions.getTriggerCompatibleActions(CONTEXT_MENU_TRIGGER, {
+          embeddable,
+        })) ?? []
+      );
+    })();
+
+    let allActions = regularActions.concat(
+      Object.values(universalActions ?? {}) as Array<Action<object>>
+    );
+
+    const { disabledActions } = embeddable.getInput();
+    if (disabledActions) {
+      const removeDisabledActions = removeById(disabledActions);
+      allActions = allActions.filter(removeDisabledActions);
+    }
+    allActions.sort(sortByOrderField);
+
+    if (actionPredicate) {
+      allActions = allActions.filter(({ id }) => actionPredicate(id));
+    }
+    return allActions;
+  }, [actionPredicate, embeddable, universalActions, getActions]);
+
+  const updatePanelActions = useCallback(async () => {
+    const newActions = await getAllPanelActions();
+    if (!mounted()) return;
+    setContextMenuActions(newActions);
+  }, [getAllPanelActions, mounted]);
+
+  /**
+   * On embeddable creation get all actions then subscribe to all
+   * input updates to refresh them
+   */
+  useEffect(() => {
+    let subscription: Subscription;
+
+    updatePanelActions().then(() => {
```

Comment:
This may cause a race condition where useEffect could run again before `updatePanelActions` returns from the first time useEffect is called. That would set up a subscription that is then never canceled.

You should protect `.then` path that does not create subscriptions if useEffect is triggered before `updatePanelActions` returns. This should do the trick

```
useEffect(() => {
  let ignore = false;
  let subscription: Subscription;

  updatePanelActions().then(() => {
    if (mounted() && !ignore) {
      /**
       * since any piece of state could theoretically change which actions are available we need to
       * recalculate them on any input change or any parent input change.
       */
      subscription = embeddable.getInput$().subscribe(() => updatePanelActions());
      if (embeddable.parent) {
        subscription.add(embeddable.parent.getInput$().subscribe(() => updatePanelActions()));
      }
    }
  });
  return () => {
    ignore = true;
    subscription?.unsubscribe();
  };
}, [embeddable, getAllPanelActions, updatePanelActions, mounted]);
```